### DISCUSSION TITLE
Add keymap to delete according to whether text is selected or not

### DISCUSSION THREAD

**Discussion by varaskkar** [helix-editor/helix#9481] [General]
Hello! I love Helix ♥ but delete things it's a bit hard for me. Would it be possible a smart way to delete things?

For example, adding 2 keymaps which contain these functionalities:

### Keymap - Backspace

- if nothing is selected, delete a character (**delete_char_backward**)
- if something is selected, deletes it (**delete_selection**)

### Keymap - Del

- if nothing is selected, delete a character (**delete_char_forward**)
- if something is selected, deletes it (**delete_selection**)

So when we press the keys, it removes the text or char. The only difference between them would be the deletion direction.

**Comment by JacobKochems** [comment]
Hello @varaskkar, you can put the following in your `~/.config/helix/config.toml`:
```
[keys.normal]
del = "delete_char_forward"
backspace = "delete_char_backward"

[keys.select]
del = "delete_selection"
backspace = "delete_selection"
```
This should do the trick.

**Reply by varaskkar** [reply]
Thank you @JacobKochems but I mean with the same key and without the need to change mode.

Let's assume we're in the Insert Mode.
You press just 1 key, backspace for example, and 1 of the 2 actions can be performed 

```
[keys.insert]
backspace = "smart_delete_backward"
del = "smart_delete_forward"
```
`smart_delete_backward` would be a new keymap would perform 2 actions, according the situation
- if nothing is selected, delete a character (**it should run -> _delete_char_backward_**)
- if something is selected, deletes it (**it should run -> _delete_selection_**)

`smart_delete_forward` would be a new keymap would perform 2 actions, according the situation
- if nothing is selected, delete a character (**it should run -> _delete_char_forward_**)
- if something is selected, deletes it (**it should run -> _delete_selection_**)

What I suggest is a bit specific and I think that by default doesn't exist in Helix, would it be possible to have something like this that works depending on the selected text?

**Comment by JacobKochems** [comment]
So, as far as I understand you, you want the type of delete action to be conditional on the selection present without explicitly changing into `select_mode`. Like, for example, when you navigate by words with `w` in `normal_mode` and you then may or may not enter `insert_mode`, hitting either backspace or delete should delete the current selection and when the selection has a width of one character it should perform a character delete in the direction associated with the pressed key (`del` or `backspace`).

I hope I got that right and it helps to clarify the use case for further discussions.

**Comment by pascalkuthe** [comment]
`d` already acts like that in normal mode. There is also no such thing as no selection. You always have atleast one caharacter selected. Just delete in normal mode. Insert mode backspace is only meant for correcting small typos not acting on selections.

**Reply by JacobKochems** [reply]
Assuming I was correct in my description of @varaskkar use case, this would also interfere with one very common workflow:
1. Navigate by word with `w` in `normal_mode`
2. Enter `insert_mode` after the selection with `a`
3. Add text
4. Making a small mistake
5. Hitting backspace to correct that mistake

If you would have the proposed feature available and configured, what would happen is that you would not only accidentally delete everything you just typed but also the previously selected word before it.

**Reply by varaskkar** [reply]
> d already acts like that in normal mode

True @pascalkuthe , it's just what I was looking for but I hoped it was possible to have something similar in insert mode.

I honestly have configured Helix to work in `insert mode`, and I only use the `normal mode` to run commands or change some settings. I know it's not the default behavior but I feel comfortable with it.

I thought of something like this:

![image](https://i.imgur.com/YM725NO.jpg)

My idea was that it would be compatible with deletion from right to left, and vice versa. While someone is in `insert mode` from the beginning, act both ways by pressing the `backspace` key.

Do you think this is a feasible idea or is it out of scope for Helix?

**Reply by the-mikedavis** [reply]
Yeah that's out of scope. Normal and select modes are meant for creating / manipulating selections while insert mode is meant to be minimal and just about basic text editing (adding text, backspace, C-w, etc.). We aren't interested in supporting multiple paradigms for using Helix, see https://github.com/helix-editor/helix/blob/cf4492174d0ee27bd3c73a5fa57fe3a26aa064be/docs/vision.md?plain=1#L11

**Reply by varaskkar** [reply]
No problem, I understand Helix's goal. Anyway, thanks for your time to see the proposal :)
